go/logic/hooks.go

```diff
@@ -0,0 +1,151 @@
+/*
+/*
+   Copyright 2016 GitHub Inc.
+	 See https://github.com/github/gh-ost/blob/master/LICENSE
+*/
+
+package logic
+
+import (
+	"fmt"
+	"os"
+	"os/exec"
+	"path/filepath"
+
+	"github.com/github/gh-ost/go/base"
+	"github.com/openark/golib/log"
+)
+
+const (
+	onStartup            = "gh-ost-on-startup"
+	onValidated          = "gh-ost-on-validated"
+	onRowCountComplete   = "gh-ost-on-rowcount-complete"
+	onBeforeRowCopy      = "gh-ost-on-before-row-copy"
+	onRowCopyComplete    = "gh-ost-on-row-copy-complete"
+	onBeginPostponed     = "gh-ost-on-begin-postponed"
+	onBeforeCutOver      = "gh-ost-on-before-cut-over"
+	onInteractiveCommand = "gh-ost-on-interactive-command"
+	onSuccess            = "gh-ost-on-success"
+	onFailure            = "gh-ost-on-failure"
+	onStatus             = "gh-ost-on-status"
+	onStopReplication    = "gh-ost-on-stop-replication"
+)
+
+type HooksExecutor struct {
+	migrationContext *base.MigrationContext
+}
+
+func NewHooksExecutor() *HooksExecutor {
+	return &HooksExecutor{
+		migrationContext: base.GetMigrationContext(),
+	}
+}
+
+func (this *HooksExecutor) initHooks() error {
+	return nil
+}
+
+func (this *HooksExecutor) applyEnvironmentVairables(extraVariables ...string) []string {
+	env := os.Environ()
+	env = append(env, fmt.Sprintf("GH_OST_DATABASE_NAME=%s", this.migrationContext.DatabaseName))
+	env = append(env, fmt.Sprintf("GH_OST_TABLE_NAME=%s", this.migrationContext.OriginalTableName))
+	env = append(env, fmt.Sprintf("GH_OST_GHOST_TABLE_NAME=%s", this.migrationContext.GetGhostTableName()))
+	env = append(env, fmt.Sprintf("GH_OST_OLD_TABLE_NAME=%s", this.migrationContext.GetOldTableName()))
+	env = append(env, fmt.Sprintf("GH_OST_DDL=%s", this.migrationContext.AlterStatement))
+	env = append(env, fmt.Sprintf("GH_OST_ELAPSED_SECONDS=%f", this.migrationContext.ElapsedTime().Seconds()))
+	env = append(env, fmt.Sprintf("GH_OST_MIGRATED_HOST=%s", this.migrationContext.ApplierConnectionConfig.ImpliedKey.Hostname))
+	env = append(env, fmt.Sprintf("GH_OST_INSPECTED_HOST=%s", this.migrationContext.InspectorConnectionConfig.ImpliedKey.Hostname))
+	env = append(env, fmt.Sprintf("GH_OST_EXECUTING_HOST=%s", this.migrationContext.Hostname))
+	env = append(env, fmt.Sprintf("GH_OST_HOOKS_HINT=%s", this.migrationContext.HooksHintMessage))
+
+	for _, variable := range extraVariables {
+		env = append(env, variable)
+	}
+	return env
+}
+
+// executeHook executes a command, and sets relevant environment variables
+func (this *HooksExecutor) executeHook(hook string, extraVariables ...string) error {
+	cmd := exec.Command(hook)
+	cmd.Env = this.applyEnvironmentVairables(extraVariables...)
+
+	if err := cmd.Run(); err != nil {
```

Comment:
I think it would be helpful to pipe hook output (stdout & stderr) to gh-ost's standard output, especially for troubleshooting hook errors.

This might be out of scope of this PR.  If so, and you are interested in the idea, I am happy to submit a PR once this is merged.